providers/hns/hns_roce_u_verbs.c

```diff
@@ -634,137 +618,153 @@ static int hns_roce_alloc_recv_inl_buf(struct ibv_qp_cap *cap,
+		size = to_hr_hem_entries_size(qp->ex_sge.sge_cnt,
+					      qp->ex_sge.sge_shift);
+		qp->buf_size += size;
 	}
 
+	/* RQ WQE */
+	rq->offset = qp->buf_size;
+	size = to_hr_hem_entries_size(rq->wqe_cnt, rq->wqe_shift);
+	qp->buf_size += size;
+
+	if (qp->buf_size < 1)
+		return EINVAL;
+
 	return 0;
 }
 
 static int hns_roce_alloc_qp_buf(struct ibv_pd *pd, struct ibv_qp_cap *cap,
-				 enum ibv_qp_type type, struct hns_roce_qp *qp)
+				 struct hns_roce_qp *qp)
 {
-	int page_size = to_hr_dev(pd->context->device)->page_size;
+	if (calc_qp_buff_size(pd, qp))
+		return EINVAL;
 
 	qp->sq.wrid = malloc(qp->sq.wqe_cnt * sizeof(uint64_t));
 	if (!qp->sq.wrid)
 		return -1;
 
 	if (qp->rq.wqe_cnt) {
 		qp->rq.wrid = malloc(qp->rq.wqe_cnt * sizeof(uint64_t));
-		if (!qp->rq.wrid) {
-			free(qp->sq.wrid);
-			return -1;
-		}
+		if (!qp->rq.wrid)
+			goto err_alloc;
 	}
 
-	if (hns_roce_calc_qp_buff_size(pd, cap, type, qp)) {
-		if (qp->rq.wqe_cnt)
-			free(qp->rq.wrid);
-		free(qp->sq.wrid);
-		return -1;
+	if (qp->rq_rinl_buf.wqe_cnt) {
+		if (hns_roce_alloc_recv_inl_buf(cap, qp))
+			goto err_alloc;
 	}
 
-	if (hns_roce_alloc_buf(&qp->buf, align(qp->buf_size, page_size),
-			       to_hr_dev(pd->context->device)->page_size)) {
-		if (qp->rq.wqe_cnt)
-			free(qp->rq.wrid);
+	if (hns_roce_alloc_buf(&qp->buf, qp->buf_size, HNS_HW_PAGE_SIZE))
+		goto err_alloc;
+
+	return 0;
+
+err_alloc:
+	hns_roce_free_recv_inl_buf(qp);
+	if (qp->rq.wrid)
+		free(qp->rq.wrid);
+
+	if (qp->sq.wrid)
 		free(qp->sq.wrid);
-		return -1;
+
+	return ENOMEM;
+}
+
+static void set_extend_sge_param(struct hns_roce_device *hr_dev,
+				 struct ibv_qp_init_attr *attr,
+				 struct hns_roce_qp *qp, unsigned int wr_cnt)
+{
+	int cnt = 0;
+
+	if (hr_dev->hw_version == HNS_ROCE_HW_VER1) {
+		qp->sq.max_gs = HNS_ROCE_SGE_IN_WQE;
+	} else {
+		qp->sq.max_gs = attr->cap.max_send_sge;
+		if (attr->qp_type == IBV_QPT_UD)
+			cnt = roundup_pow_of_two(wr_cnt * qp->sq.max_gs);
+		else if (qp->sq.max_gs > HNS_ROCE_SGE_IN_WQE)
+			cnt = roundup_pow_of_two(wr_cnt *
+						 (qp->sq.max_gs -
+						  HNS_ROCE_SGE_IN_WQE));
 	}
 
-	return 0;
+	qp->ex_sge.sge_shift = HNS_ROCE_SGE_SHIFT;
+	qp->ex_sge.sge_cnt = cnt;
 }
 
 static void hns_roce_set_qp_params(struct ibv_pd *pd,
 				   struct ibv_qp_init_attr *attr,
 				   struct hns_roce_qp *qp,
 				   struct hns_roce_context *ctx)
 {
-	unsigned int sge_ex_count;
+	struct hns_roce_device *hr_dev = to_hr_dev(pd->context->device);
+	unsigned int cnt;
 
-	if (to_hr_dev(pd->context->device)->hw_version == HNS_ROCE_HW_VER1) {
-		qp->sq.wqe_cnt = align_qp_size(attr->cap.max_send_wr);
-		qp->rq.wqe_cnt = align_qp_size(attr->cap.max_recv_wr);
-	} else {
-		qp->sq.wqe_cnt = align_queue_size(attr->cap.max_send_wr);
-		qp->rq.wqe_cnt = align_queue_size(attr->cap.max_recv_wr);
-	}
+	/* RQ WQE */
+	qp->rq.max_gs = max(1U, attr->cap.max_recv_sge);
+	if (hr_dev->hw_version == HNS_ROCE_HW_VER1)
+		qp->rq.wqe_shift = hr_ilog32(sizeof(struct hns_roce_rc_rq_wqe));
```

Comment:
I see that I missed the commit 3ebc62089441860367c0534fb40832a1c9d251b6.
All those hr_ilog32() should be converted to be like in this patch 69dbf7ccd18af7c7dd18a16779b03fbce59def45

Reply:
Thanks for your reminder, but I have fixed it in 3ebc620 :)